components/proxy/src/main/java/com/hotels/styx/routing/handlers/PathPrefixRouter.java

```diff
@@ -0,0 +1,121 @@
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx.routing.handlers;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.hotels.styx.api.Eventual;
+import com.hotels.styx.api.HttpHandler;
+import com.hotels.styx.api.LiveHttpRequest;
+import com.hotels.styx.common.Pair;
+import com.hotels.styx.config.schema.Schema;
+import com.hotels.styx.infrastructure.configuration.yaml.JsonNodeConfig;
+import com.hotels.styx.routing.config.HttpHandlerFactory;
+import com.hotels.styx.routing.config.RoutingObjectDefinition;
+import com.hotels.styx.server.NoServiceConfiguredException;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentSkipListMap;
+
+import static com.hotels.styx.common.Pair.pair;
+import static com.hotels.styx.config.schema.SchemaDsl.field;
+import static com.hotels.styx.config.schema.SchemaDsl.list;
+import static com.hotels.styx.config.schema.SchemaDsl.object;
+import static com.hotels.styx.config.schema.SchemaDsl.routingObject;
+import static com.hotels.styx.config.schema.SchemaDsl.string;
+import static com.hotels.styx.routing.config.RoutingConfigParser.toRoutingConfigNode;
+import static com.hotels.styx.routing.config.RoutingSupport.missingAttributeError;
+import static java.lang.String.join;
+import static java.util.Comparator.comparingInt;
+import static java.util.Comparator.naturalOrder;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Makes a routing decision based on a request path prefix.
+ *
+ * Chooses a destination according to longest matching path prefix.
+ * The destination can be a routing object reference or an inline definition.
+ */
+public class PathPrefixRouter {
+    public static final Schema.FieldType SCHEMA = object(
+            field("routes", list(object(
+                    field("prefix", string()),
+                    field("destination", routingObject()))
+            ))
+    );
+
+    private final ConcurrentSkipListMap<String, HttpHandler> routes = new ConcurrentSkipListMap<>(
+                    comparingInt(String::length)
+                            .reversed()
+                            .thenComparing(naturalOrder())
+    );
+
+    PathPrefixRouter(List<Pair<String, HttpHandler>> routes) {
+        routes.forEach(entry -> this.routes.put(entry.key(), entry.value()));
+    }
+
+    public Optional<HttpHandler> route(LiveHttpRequest request) {
+        String path = request.path();
+
+        return routes.entrySet().stream()
+                .filter(entry -> path.startsWith(entry.getKey()))
+                .findFirst()
+                .map(Map.Entry::getValue);
+    }
+
+    public static class Factory implements HttpHandlerFactory {
```

Comment:
It is to simplify unit testing.

Otherwise `PathPrefixRouter` would have to be a `HttpHandler` with a return type of `Eventual<LiveHttpResponse>`. This would complicate unit testing.

The `Factory` for every class in `routing.handlers.*` returns a routing object as a `HttpHandler` instance. From the consumer's point of view it doesn't matter if `PathPrefixRouter` is the underlying implementation, or merely just a part of it. But it being merely an internal component of an `HttpHandler` simplifies unit tests and the divergence is preferred.